core/metadata/content.go

```diff
@@ -574,6 +574,13 @@ func (nw *namespacedWriter) Commit(ctx context.Context, size int64, expected dig
 
 	var innerErr error
 
+	// We pre-sync the in-flight writes to the disk.  This avoids the [subsequent fp.Sync() call]
+	// (https://github.com/containerd/containerd/blob/c4c3c6ea568ce0cfbcf754863abadeea37d77c8f/plugins/content/local/writer.go#L95)
+	// from taking too long (10s+) while holding the metadata database lock as in the following
+	// `update` transaction.  We intentionally ignore any error on Sync() because it will be
+	// handled by the subsequent `fp.Sync` anyway.
+	nw.Sync()
```

Comment:
The idea looks good. However, as far as I know, fsync or fdatasync could commit unrelated dirty page. 
When you do it outside, it could impact other sync syscall, like NewContainer, because bbolt always need to call fdatasync for commit. If you can provide some performance test result, it would be better. Thanks

REF: https://lwn.net/Articles/842385/

Reply:
@fuweid Thanks for reviewing the PR!

> When you do it outside, it could impact other sync syscall

This applies regardless of where do the sync, right?  The downside of doing the sync under an exclusive transaction/lock is that it blocks concurrent operations that don't demand heavy I/O like `CreateContainer`.

> because bbolt always need to call fdatasync for commit

That fdatasync call is on the metadata database, not on the image layers (which can be gigabytes of data).  So the degree of lock contention is dramatically different.

> If you can provide some performance test result, it would be better.

In our production environment, we pull and start around 10 containers upon VM booting.  We observed a high variance in `CreateContainer` gRPCs, and sometimes a container creation can be blocked by over a minute.  With this change, the latency of `CreateContainer` becomes stable and negligible (finish in a second or two).

Reply:
Thanks for the comment. 

> This applies regardless of where we do the sync, right?

Yes. I was saying that the CreateContainer can commit part of dirty pages created by content writer. 
It's random based on the ext4 fast commit article.

> That fdatasync call is on the metadata database, not on the image layers (which can be gigabytes of data). So the degree of lock contention is dramatically different.

Please checkout filesystem behaviour I mentioned in the REF. Both metadata database and image layers are in the same block device by default. They can impact each other.

> In our production environment, we pull and start around 10 containers upon VM booting. We observed a high variance in CreateContainer gRPCs, and sometimes a container creation can be blocked by over a minute. With this change, the latency of CreateContainer becomes both stable and negligible (e.g., in a second or two).

That's why I want to see that improvement result. It will be more confident to get this commit.

Reply:
Makes sense.  We've seen a substantial improvement on the `CreateContainer` latency in our setup.

Reply:
> Please checkout filesystem behaviour I mentioned in the REF. Both metadata database and image layers are in the same block device by default. They can impact each other.

@fuweid My understanding is that even if this exists today, we are still better off moving the sync outside of the lock which is almost guaranteed to be blocking should there be a large image layer (e.g., GB+).  And there's also a chance the interference issue you referenced will be addressed by the linux kernel community in the future, and then we are having a free ride here :)